### ISSUE TITLE
BUG: Mixed-precision configuration not working with STATIC quantization

### ISSUE THREAD

**Issue by sasha-hailo** [ModelTC/llmc#163]
Dear LLMC team,
I've been trying to run mixed-precision PTQ quantization using RTN.
I suspect there's a bug, as the **non-default settings in `mix_bits` are ignored**.

My understanding of the code:
- In method `get_act_qparams()` of `rtn.py`, the values of `qmax` / `qmin` / `scales` / `zeros` are determined using the **default** quantizer bit precision
- These values are registered as `buf_act_<xxx>` buffers, _for all modules / layers_.
- During inference time, in method `a_qdq()` of `rtn.py`, though the `aquantizer` object of each layer is configured correctly, it blindly loads from buffer the registered quantization parameters `qmin` / `qmax` / `scales` / `zeros`, and uses them, instead of the **actual** values it should support.

What do you think?
Thanks in advance!

**Comment by Harahan** [comment]
There's no ``get_act_qparams()`` in ``rtn.py``. You can print the bit-width of each linear to check the code.

**Comment by sasha-hailo** [comment]
Hi @Harahan,
Thank you for your response.
It turns out that a lot of changes have been made since my issue report ([in this commit](https://github.com/ModelTC/llmc/commit/3a6b62ebbff4f9e4b442fa677c9af24a396ff1b6)).
The functionality I was referring to as `get_act_qparams()` now resides in `register_act_qparams()` in file `base_blockwise_quantization.py`.

**The bug, unfortunately, persists.**

The "mechanism" is the same: function `register_act_qparams()` uses a single quantizer object (`self.aquantizer`) to determine the quantization parameters of all layers - and this quantizer is configured with the _default_ settings. It determines the scale & zero point settings (**w.r.t. incorrect bit width**), and registers them via `buf_act_scales` / `buf_act_zeros`.

Note that the correct per-layer quantization configurations are loaded when executing `deploy()` function, 
But they have no effect, as they are using the incorrect scale & zero-point values determined in the previous stage!

To sum it up: I think that the core issue that causes the [suspected] bug is that the calibration stage & function `register_act_qparams()` are unaware of the configured mixed-precision, and work with the default quantization config.
This code could work well in _dynamic_ quantization, but not in a _static_ quantization scenario.
I also suspect that the same issue can happen with other quantization methods.

Can you please look into it?
Thanks in advance!

**Comment by sasha-hailo** [comment]
P.S.
An unrelated question:
I also noticed that the commit I mentioned above added some limited support to additional quantization granularity, via functions `get_matmul_in_block()`, `get_softmax_in_block()`, `get_act_fn_in_block()`.
Do you plan to extend this support to the more common LLM models like Qwen & LLama?
(This could be really cool)

**Comment by Harahan** [comment]
It depends on whether we encounter such a need or it will be used in our research. So, not sure.

**Comment by sasha-hailo** [comment]
Did you succeed in reproducing the `mix_bits` problem I reported?
I believe the issue should be reopened as a bug...

**Comment by Harahan** [comment]
I'm sorry, but we do not have enough time to do this. If you are sure there's a bug, post the error log and reopen the issue.

**Comment by sasha-hailo** [comment]
[LLMC_RTN_W8A8_MixedA16_Bug.txt](https://github.com/user-attachments/files/17635207/LLMC_RTN_W8A8_MixedA16_Bug.txt)
[LLMC_RTN_W8A8.txt](https://github.com/user-attachments/files/17635208/LLMC_RTN_W8A8.txt)

I'm pretty sure this is a bug.
And I now suspect that the issue affects not only RTN, but nearly any method based on *static* quantization.
Can you please reopen the issue? I don't think I have the permissions for this.

Please find attached 2 logs of LLMC with an RTN configuration.
One log refers to a configuration without mix-bit, the other with mix-bit.
If you compare the two files, you can see that
- The outputs of both runs are identical (same PPL score), hinting that the `mix_bit` configuration had no effect.
- The `mix_bit` configuration of the deployed model is correct (see circa line 2458 in the log) 
==> the bug is not at the deployment stage, but at the calibration stage (see my explanation in earlier messages).

**Comment by Harahan** [comment]
I reopen the issue. Since we currently don't have the requirement for the static quantization, the bug may be fixed a long time later.

**Comment by nelaturuharsha** [comment]
Hi,

I wanted to start using this library for a couple of things, but just to confirm, this bug affects situations where:
Static quantization is applied layer-wise (with the intention to have different layers/components at different bit-widths)

Can it be confirmed that it does not apply when I would like to more or less have the same bit-width for all components of the model or different for activations/weights?

**Comment by sasha-hailo** [comment]
To the best of my understanding,
If the quantization configuration is the same for all layers of the model, the bug does not apply.

**Comment by sasha-hailo** [comment]
Hi @Harahan , LLMC folks,
I wanted to let you know that I have solved the bug in my side branch.
In addition, I also added support of separate configurable quantization of activation _outputs_ [currently, only for linear layers].

If you're interested in any of these, please let me know, and I'll be glad to share my code or open a PR.
Note, though, that the changes are quite extensive and will require time and commitment from your side to review.

**Comment by AaronMaYue** [comment]
> Hi [@Harahan](https://github.com/Harahan) , LLMC folks, I wanted to let you know that I have solved the bug in my side branch. In addition, I also added support of separate configurable quantization of activation _outputs_ [currently, only for linear layers].
> 
> If you're interested in any of these, please let me know, and I'll be glad to share my code or open a PR. Note, though, that the changes are quite extensive and will require time and commitment from your side to review.

Hi, @sasha-hailo, I was using smoothquant's mixed bits quantization and ran into the same bug. How did you solve it? Please share your code.

**Comment by sasha-hailo** [comment]
@AaronMaYue , I apologize for the late response.
If it's still relevant, please let me know and I'll clean up my code for sharing.

**Comment by AaronMaYue** [comment]
> [@AaronMaYue](https://github.com/AaronMaYue) , I apologize for the late response. If it's still relevant, please let me know and I'll clean up my code for sharing.

Yeah, it's still relevant. The mix-precision problem remains, too. looking forward to your sharing.

**Comment by sasha-hailo** [comment]
@AaronMaYue ,
I organized and pushed my modifications to https://github.com/sasha-hailo/llmc/tree/main_hailo_share.
Note: I never tested it with SmoothQuant. Hope it works for you.

@Harahan , @gushiqiao - please consider using my code to fix the bug (as well as many additional issues).
I'll be glad to assist, if needed.

**Comment by AaronMaYue** [comment]
> [@AaronMaYue](https://github.com/AaronMaYue) , I organized and pushed my modifications to https://github.com/sasha-hailo/llmc/tree/main_hailo_share. Note: I never tested it with SmoothQuant. Hope it works for you.
> 
> [@Harahan](https://github.com/Harahan) , [@gushiqiao](https://github.com/gushiqiao) - would you like to consider using my code to fix the bug (as well as many additional issues)? I'll be glad to assist, if needed.

Hi,@sasha-hailo, thank you very much for your work, your code works😊, I tested it on both Smoothquant and RTN, the bug fixed.

**Comment by wangddcsu-ui** [comment]
最新的代码不支持层间混合量化功能里吗？即下面这种配置：
quant:
    method: Awq
    weight:
        bit: 8
        symmetric: True
        granularity: per_channel
    mix_bits:
        setting_0:
            layer_name: [down_proj]
            do_quant: True
            weight:
                bit: 4
                symmetric: True
                granularity: per_channel
我在代码的py文件中已经找不到mix_bits相关的代码，也就是说此功能废弃了吗？
目前只有ignored_layers:跳过量化的功能存在了。
